### PR TITLE
Block skins

### PR DESCRIPTION
**Pull request by msoechting**

### CODE REVIEW

File: src/tetris/brick/brick.ts [hunk 1]
```diff
@@ -28,6 +28,14 @@ export default class Brick {
 	public set stateChanged(changed: boolean) {
 		this._stateChanged = changed;
 	}
+	
+	public get custom() {
```
Comment (julianweise) [inline src/tetris/brick/brick.ts]: return value is missing

File: src/tetris/brick/brickFactory.ts [hunk 2]
```diff
@@ -19,16 +21,22 @@ export default class BrickFactory {
 
 	//region public methods
 	public newBrick(field: Field): Brick {
-		const blockAssetId = this._selectBlockAssetId();
 		const bias = this._biasEngine.newBrickBias(field);
-		return this._newBrick(blockAssetId, bias.position, bias.chances, field);
+		return this._newBrick(bias.position, bias.chances, field);
+	}
+
+	public newCustomBrick(brickType: BrickType, position: Vector2) {
```
Comment (julianweise) [inline src/tetris/brick/brickFactory.ts]: return value is missing

File: src/tetris/brick/brickType.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+enum BrickType {
+    'I', 'O', 'J', 'L', 'T', 'S', 'Z'
```
Comment (julianweise) [inline src/tetris/brick/brickType.ts]: Usually, enum initialization should be done without quotes

File: src/tetris/brick/skin.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import { SkinRarity } from "tetris/brick/skinRarity";
```
Comment (julianweise) [inline src/tetris/brick/skin.ts]: Regions are missing

File: src/tetris/brick/skin.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import BrickType from "tetris/brick/brickType";
+import Brick from "tetris/brick/brick";
+
+export default class Skin {
+    private _frameName: string;
```
Comment (julianweise) [inline src/tetris/brick/skin.ts]: readonly

File: src/tetris/brick/skin.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    private _rarity: SkinRarity;
```
Comment (julianweise) [inline src/tetris/brick/skin.ts]: readonly

File: src/tetris/brick/skin.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    private _brickType: BrickType;
```
Comment (julianweise) [inline src/tetris/brick/skin.ts]: readonly

File: src/tetris/brick/skin.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    private _unlocked: boolean = false;
+    private _name: string;
```
Comment (julianweise) [inline src/tetris/brick/skin.ts]: readonly

File: src/tetris/brick/skin.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    private _id: number;
```
Comment (julianweise) [inline src/tetris/brick/skin.ts]: readonly

File: src/tetris/brick/skin.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    public constructor(frameName: string, rarity: SkinRarity, brickType: BrickType, name: string, id: number) {
+		this._frameName = frameName;
+        this._rarity = rarity;
+        this._brickType = brickType;
+        this._name = name;
+        this._id = id;
+    }
+    
+    public unlock() {
```
Comment (julianweise) [inline src/tetris/brick/skin.ts]: return type is missing

File: src/tetris/brick/skinStorage.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import BrickType from "tetris/brick/brickType";
```
Comment (julianweise) [inline src/tetris/brick/skinStorage.ts]: regions are missing

File: src/tetris/brick/skinStorage.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import Skin from "tetris/brick/skin";
+import { SkinRarity } from "tetris/brick/skinRarity";
+
+export default class SkinStorage {
+    private skins: Map<BrickType, Array<Skin>>;
```
Comment (julianweise) [inline src/tetris/brick/skinStorage.ts]: readonly

File: src/tetris/brick/skinStorage.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    private equippedSkins: Map<BrickType, Skin>;
```
Comment (julianweise) [inline src/tetris/brick/skinStorage.ts]: readonly

File: src/tetris/brick/skinStorage.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    public skinAmount: number;
```
Comment (julianweise) [inline src/tetris/brick/skinStorage.ts]: readonly

File: src/tetris/brick/skinStorage.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+    public constructor() {
+        this.skins = new Map<BrickType, Array<Skin>>();
+        this.equippedSkins = new Map<BrickType, Skin>();
+        const skinNameBases = ["colorful", "bw", "sepia", /* rare */ "aqua", "light", "purple", /* epic*/ "flat", "glitch", "shiny", /* legendary*/ "dotted", "relief", "retro",];
```
Comment (julianweise) [inline src/tetris/brick/skinStorage.ts]: line too long

File: src/tetris/brick/skinStorage.ts [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+        for (let brickType = 0; brickType < 7; brickType++) {
```
Comment (julianweise) [inline src/tetris/brick/skinStorage.ts]: magic number, you could use `Object.keys(ENUM).length`

File: src/tetris/brick/skinStorage.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+            const brickSkins = [];
+            for (let s = 0; s < skinNameBases.length; s++) {
+                for (let i = 1; i < 7; i++) {
+                    const rarity = Math.floor(s / 3);
+                    const name = [skinNameBases[s], i].join(' ');
+                    brickSkins.push(new Skin(skinNameBases[s] + '_0' + i, rarity, brickType, name, brickSkins.length));
```
Comment (julianweise) [inline src/tetris/brick/skinStorage.ts]: line too long.
You could introduce a new function for initializing a skin

File: src/tetris/brick/skin.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
+import { SkinRarity } from "tetris/brick/skinRarity";
+import BrickType from "tetris/brick/brickType";
+import Brick from "tetris/brick/brick";
+
+export default class Skin {
```
Comment (Johannes-Schneider) [inline src/tetris/brick/skin.ts]: Please use our trusty region pattern.

File: src/tetris/brick/skinStorage.ts [hunk 19]
```diff
@@ -0,0 +1,41 @@
+import BrickType from "tetris/brick/brickType";
+import Skin from "tetris/brick/skin";
+import { SkinRarity } from "tetris/brick/skinRarity";
+
+export default class SkinStorage {
```
Comment (Johannes-Schneider) [inline src/tetris/brick/skinStorage.ts]: Please use our trusty region pattern.

File: src/tetris/scene/collectionScene.ts [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	public update(time: number, delta: number): void {
+		this._pipeline.setFloat1('uTime', time / 2600);
+	}
+	//endregion
+
+	//region constructor
+	public constructor(game: Game, skinStorage: SkinStorage) {
+		super({
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: object with one property should be on one line

File: src/tetris/scene/collectionScene.ts [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+			key: "CollectionScene"
+		});
+		this._game = game;
+		this._skinStorage = skinStorage;
+		this._selectedSkins = new Map<BrickType, Skin>();
+
+		for (let brickType = 0; brickType < 7; brickType++) {
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: magic number (7). See comment above

File: src/tetris/scene/collectionScene.ts [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	private _createButtons(): void {
+		const menuX1: number = config.graphics.width / 3;
+		const menuX2: number = 2 * config.graphics.width / 3;
+		const menuY: number = config.graphics.height / 5;
+		const spacingX: number = 100;
+		const spacingY: number = config.graphics.height / 6;
+
+		new TextButton(this, config.graphics.width / 2, config.graphics.height * 7 / 8, "green_button00.png", "green_button01.png", "OK", function(){ this._game.changeScene(config.sceneKeys.menuScene); }.bind(this));
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: line too long

File: src/tetris/scene/collectionScene.ts [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+	//endregion
+
+	//region public methods
+	public create(): void {
+		this._createBackground();
+		this._createButtons();
+	}
+
+	public update(time: number, delta: number): void {
+		this._pipeline.setFloat1('uTime', time / 2600);
```
Comment (Johannes-Schneider) [inline src/tetris/scene/collectionScene.ts]: Why `time / 2600`?
Seems not to be a time conversion.

File: src/tetris/scene/collectionScene.ts [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+
+		this._background.setPipeline('rainbow');
+	}
+
+	private _createButtons(): void {
+		const menuX1: number = config.graphics.width / 3;
+		const menuX2: number = 2 * config.graphics.width / 3;
+		const menuY: number = config.graphics.height / 5;
+		const spacingX: number = 100;
+		const spacingY: number = config.graphics.height / 6;
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: Lots of magic numbers (3, 5, 6, 7, 8) in line 76, 77, 78, 80, 82

File: src/tetris/scene/collectionScene.ts [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+		new TextButton(this, config.graphics.width / 2, config.graphics.height * 7 / 8, "green_button00.png", "green_button01.png", "OK", function(){ this._game.changeScene(config.sceneKeys.menuScene); }.bind(this));
+		
+		const customBrickOffsets = [new Vector2(-0.5, 0), new Vector2(-1, -1), new Vector2(-1, -0.5), new Vector2(0, -0.5), new Vector2(-1.5, -1), new Vector2(-1.5, -1), new Vector2(-1.5, -1) ];
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: line too long

File: src/tetris/scene/collectionScene.ts [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+
+		for (let i = 0; i < 7; i++) {
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: lots of magic numbers in for loop

File: src/tetris/scene/collectionScene.ts [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+			const x = i < 4 ? menuX1 : menuX2;
+			const y = menuY + (i % 4) * spacingY + (i / 4) * spacingY * 0.5;
+			const brick = new BrickFactory(this, null, null).newCustomBrick(i as BrickType, customBrickOffsets[i]);
+			const text = this.add.bitmapText(0, y + 50, config.ui.fontKeys.kenneyMiniSquare, this._selectedSkins.get(i).name);
+			text.x = x - text.width / 2;
+			new TextButton(this, x - spacingX, y, "blue_sliderLeft.png", "blue_sliderLeft.png", "", this._changeSkin.bind(this, brick, i, -1, text, x));
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: line too long

File: src/tetris/scene/collectionScene.ts [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+			brick.preDraw(new Vector2(x, y));
+			new TextButton(this, x + spacingX, y, "blue_sliderRight.png", "blue_sliderRight.png", "", this._changeSkin.bind(this, brick, i, 1, text, x));
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: line too long

File: src/tetris/scene/collectionScene.ts [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+		}
+	}
+
+	private _changeSkin(brick: Brick, brickType: BrickType, indexMovement: number, text: Phaser.GameObjects.BitmapText, textBaseX: number) {
+		const newSkin = this._skinStorage.getSkin(brickType, (this._selectedSkins.get(brickType).id + indexMovement + this._skinStorage.skinAmount) % this._skinStorage.skinAmount);
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: line too long

File: src/tetris/scene/collectionScene.ts [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+			const brick = new BrickFactory(this, null, null).newCustomBrick(i as BrickType, customBrickOffsets[i]);
+			const text = this.add.bitmapText(0, y + 50, config.ui.fontKeys.kenneyMiniSquare, this._selectedSkins.get(i).name);
+			text.x = x - text.width / 2;
+			new TextButton(this, x - spacingX, y, "blue_sliderLeft.png", "blue_sliderLeft.png", "", this._changeSkin.bind(this, brick, i, -1, text, x));
+			brick.preDraw(new Vector2(x, y));
+			new TextButton(this, x + spacingX, y, "blue_sliderRight.png", "blue_sliderRight.png", "", this._changeSkin.bind(this, brick, i, 1, text, x));
+		}
+	}
+
+	private _changeSkin(brick: Brick, brickType: BrickType, indexMovement: number, text: Phaser.GameObjects.BitmapText, textBaseX: number) {
```
Comment (julianweise) [inline src/tetris/scene/collectionScene.ts]: missing return type

File: src/tetris/scene/menuScene.ts [hunk 31]
```diff
@@ -16,8 +16,12 @@ export default class MenuScene extends Phaser.Scene {
 
 	//region public methods
 	public preload(): void {
+		this.load.atlas(config.atlasKeys.blockSpriteAtlasKey, "./assets/images/blockSprites.png", "./assets/images/blockSprites.json");
```
Comment (julianweise) [inline src/tetris/scene/menuScene.ts]: line too long

File: src/tetris/scene/menuScene.ts [hunk 32]
```diff
@@ -16,8 +16,12 @@ export default class MenuScene extends Phaser.Scene {
...
 		this.load.atlas(config.atlasKeys.uiSpriteAtlasKey, "./assets/images/uiSprites.png", "./assets/images/uiSprites.json");
-		this.load.glsl('rainbow', "./assets/shaders/rainbow.glsl")
+		this.load.image(config.graphics.noiseTextureKey, "./assets/images/noise.png");
+		this.load.bitmapFont(config.ui.fontKeys.kenneyMiniSquare, "./assets/font/KenneyMiniSquare.png", "./assets/font/KenneyMiniSquare.fnt", );
```
Comment (julianweise) [inline src/tetris/scene/menuScene.ts]: line too long

File: src/tetris/scene/menuScene.ts [hunk 33]
```diff
@@ -69,31 +73,31 @@ export default class MenuScene extends Phaser.Scene {
...
-		const buttonsXPosition: number = config.graphics.width / 2;
+		const buttonPositionX: number = config.graphics.width / 2;
 		const menuStartY: number = config.graphics.height / 3;
 		const spacing: number = 20;
 
-		this._playButton = new TextButton(this, buttonsXPosition, 0, "blue_button00.png", "blue_button01.png", "Join Matchmaking", this._joinMatchmaking.bind(this));
-		this._optionsButton = new TextButton(this, buttonsXPosition, 0, "blue_button00.png", "blue_button01.png", "Leave Matchmaking", this._leaveMatchmaking.bind(this));
-
+		this._playButton = new TextButton(this, buttonPositionX, 0, "green_button00.png", "green_button01.png", "Quick Match", this._joinMatchmaking.bind(this));
+		this._collectionButton = new TextButton(this, buttonPositionX, 0, "blue_button00.png", "blue_button01.png", "My Collection", function(){ this._game.changeScene(config.sceneKeys.collectionScene); }.bind(this) );
```
Comment (julianweise) [inline src/tetris/scene/menuScene.ts]: line too long

File: src/tetris/scene/menuScene.ts [hunk 34]
```diff
@@ -69,31 +73,31 @@ export default class MenuScene extends Phaser.Scene {
...
 	private _createButtons(): void {
-		const buttonsXPosition: number = config.graphics.width / 2;
+		const buttonPositionX: number = config.graphics.width / 2;
 		const menuStartY: number = config.graphics.height / 3;
 		const spacing: number = 20;
 
-		this._playButton = new TextButton(this, buttonsXPosition, 0, "blue_button00.png", "blue_button01.png", "Join Matchmaking", this._joinMatchmaking.bind(this));
-		this._optionsButton = new TextButton(this, buttonsXPosition, 0, "blue_button00.png", "blue_button01.png", "Leave Matchmaking", this._leaveMatchmaking.bind(this));
-
+		this._playButton = new TextButton(this, buttonPositionX, 0, "green_button00.png", "green_button01.png", "Quick Match", this._joinMatchmaking.bind(this));
```
Comment (julianweise) [inline src/tetris/scene/menuScene.ts]: line too long

File: src/tetris/scene/collectionScene.ts (hunk 23 above)
Comment (msoechting) [reply src/tetris/scene/collectionScene.ts]: ✨ its magic

File: src/tetris/brick/skin.ts (hunk 18 above)
Comment (msoechting) [reply src/tetris/brick/skin.ts]: done

File: src/tetris/brick/skinStorage.ts (hunk 19 above)
Comment (msoechting) [reply src/tetris/brick/skinStorage.ts]: done

File: src/tetris/brick/brick.ts (hunk 1 above)
Comment (msoechting) [reply src/tetris/brick/brick.ts]: done

File: src/tetris/brick/brickFactory.ts (hunk 2 above)
Comment (msoechting) [reply src/tetris/brick/brickFactory.ts]: done

File: src/tetris/brick/brickType.ts (hunk 3 above)
Comment (msoechting) [reply src/tetris/brick/brickType.ts]: done

File: src/tetris/brick/skin.ts (hunk 4 above)
Comment (msoechting) [reply src/tetris/brick/skin.ts]: done

File: src/tetris/brick/skin.ts (hunk 5 above)
Comment (msoechting) [reply src/tetris/brick/skin.ts]: done

File: src/tetris/brick/skin.ts (hunk 6 above)
Comment (msoechting) [reply src/tetris/brick/skin.ts]: done

File: src/tetris/brick/skin.ts (hunk 7 above)
Comment (msoechting) [reply src/tetris/brick/skin.ts]: done

File: src/tetris/brick/skin.ts (hunk 8 above)
Comment (msoechting) [reply src/tetris/brick/skin.ts]: done

File: src/tetris/brick/skin.ts (hunk 9 above)
Comment (msoechting) [reply src/tetris/brick/skin.ts]: done

File: src/tetris/brick/skin.ts (hunk 10 above)
Comment (msoechting) [reply src/tetris/brick/skin.ts]: done

File: src/tetris/brick/skinStorage.ts (hunk 11 above)
Comment (msoechting) [reply src/tetris/brick/skinStorage.ts]: done

File: src/tetris/brick/skinStorage.ts (hunk 12 above)
Comment (msoechting) [reply src/tetris/brick/skinStorage.ts]: done

File: src/tetris/brick/skinStorage.ts (hunk 13 above)
Comment (msoechting) [reply src/tetris/brick/skinStorage.ts]: done

File: src/tetris/brick/skinStorage.ts (hunk 14 above)
Comment (msoechting) [reply src/tetris/brick/skinStorage.ts]: done

File: src/tetris/brick/skinStorage.ts (hunk 15 above)
Comment (msoechting) [reply src/tetris/brick/skinStorage.ts]: done

File: src/tetris/brick/skinStorage.ts (hunk 16 above)
Comment (msoechting) [reply src/tetris/brick/skinStorage.ts]: done

File: src/tetris/brick/skinStorage.ts [hunk 35]
```diff
@@ -1,41 +1,61 @@
...
         this.skins = new Map<BrickType, Array<Skin>>();
         this.equippedSkins = new Map<BrickType, Skin>();
-        const skinNameBases = ["colorful", "bw", "sepia", /* rare */ "aqua", "light", "purple", /* epic*/ "flat", "glitch", "shiny", /* legendary*/ "dotted", "relief", "retro",];
 
-        for (let brickType = 0; brickType < 7; brickType++) {
+        for (let brickType = 0; brickType < Object.keys(BrickType).length / 2; brickType++) {
             const brickSkins = [];
-            for (let s = 0; s < skinNameBases.length; s++) {
+            for (let s = 0; s < this.skinNameBases.length; s++) {
                 for (let i = 1; i < 7; i++) {
```
Comment (Johannes-Schneider) [inline src/tetris/brick/skinStorage.ts]: Please replace that hardcoded `7`.

**Pull request merged**